Add logger to Weights & Biases

borisdayma opened the pull request:
Add support for logging through Weights & Biases.

`WandbLogger` automatically logs metrics, model topology, gradients and best trained model.

Sample run: https://app.wandb.ai/borisd13/skorch/runs/1vs6725x?workspace=user-borisd13
You can switch tabs on the left and see computer resources, model graph, logged files, etc

borisdayma commented:
I'm now working on adding tests. Feel free if you have any comments.

BenjaminBossan commented:
Thanks for this PR.

You've probably seen the `Tensorboard` and `NeptuneLogger` callbacks. There we adopt a pattern of passing the actual logger instance to the callback (e.g. the `SummaryWriter`) instead of using the `try: import ...` pattern inside the callback. Would it be possible to change `WandbLogger` in a similar fashion?

borisdayma commented:
Yes of course, I'll make the changes.

BenjaminBossan commented on skorch/callbacks/logging.py:
```diff
@@ -205,6 +207,76 @@ def on_train_end(self, net, **kwargs):
+    """Logs best model and metrics to `Weights & Biases <https://docs.wandb.com/>`_
+
+    "Use this callback to automatically log best trained model and all metrics from
+    your net's history to Weights & Biases after each epoch.
```
I think you should use the docstring to help skorch users who are unfamiliar with W&B to get started quickly. E.g., you could specify what package they must install for this to work, i.e. a pip (or conda) instruction. You should also indicate what kind of setup they need to make beforehand (say, starting a local server). 

For a nice example, look at the docstring for [NeptuneLogger](https://skorch.readthedocs.io/en/latest/callbacks.html#skorch.callbacks.NeptuneLogger).

BenjaminBossan commented on skorch/callbacks/logging.py:
```diff
@@ -205,6 +207,76 @@ def on_train_end(self, net, **kwargs):
+    >>> import wandb
+    >>> from skorch.callbacks import WandbLogger
+    >>> wandb_run = wandb.init()
+    >>> wandb.config.update({"learning rate": 1e-3, "batch size": 32})  # optional
```
Could you indicate what this config update does?

BenjaminBossan commented on skorch/callbacks/logging.py:
```diff
@@ -205,6 +207,76 @@ def on_train_end(self, net, **kwargs):
+    """
+
+    # Record if watch has been called previously (even in another instance)
+    _watch_called = False
```
Is this really used anywhere? If it is, please move this inside `initialize` and call it `watch_called_`.

BenjaminBossan commented on skorch/callbacks/logging.py:
```diff
@@ -205,6 +207,76 @@ def on_train_end(self, net, **kwargs):
+        self.wandb_run = wandb_run
+        self.save_model = save_model
+        self.keys_ignored = keys_ignored
+        self.model_path = Path(wandb_run.dir) / 'best_model.pth'
```
Please don't set any arguments in `__init__` that are not passed by the user. So either allow them to pass the `model_path` argument (if that makes any sense) or instead set the `model_path` inside `initialize` (and call it `model_path_`).

BenjaminBossan commented on skorch/callbacks/logging.py:
```diff
@@ -205,6 +207,76 @@ def on_train_end(self, net, **kwargs):
+        """Automatically log values from the last history step."""
+        hist = net.history[-1]
+        keys_kept = filter_log_keys(hist, keys_ignored=self.keys_ignored_)
+        logged_vals = dict((k, hist[k]) for k in keys_kept if k in hist)
```
```suggestion
        logged_vals = {k: hist[k] for k in keys_kept if k in hist}
```

BenjaminBossan commented on skorch/callbacks/logging.py:
```diff
@@ -205,6 +207,76 @@ def on_train_end(self, net, **kwargs):
+      wandb Run used to log data.
+
+    save_model : bool (default=True)
+      Saves best trained model.
```
```suggestion
      Whether to save a checkpoint of the best model.
```

BenjaminBossan commented on skorch/callbacks/logging.py:
```diff
@@ -205,6 +207,76 @@ def on_train_end(self, net, **kwargs):
+    def __init__(
+            self,
+            wandb_run,
+            save_model=True,
```
We already provide a [checkpoint callback](https://skorch.readthedocs.io/en/stable/callbacks.html#skorch.callbacks.Checkpoint), I think this functionality is redundant.

borisdayma replied:
This is to log the trained model to W&B.

BenjaminBossan replied:
Interesting. How does that work? In this code, I don't see any interaction with W&B:

```python
        # save best model
        if self.save_model and hist['valid_loss_best']:
            model_path = Path(self.wandb_run.dir) / 'best_model.pth'
            with model_path.open('wb') as model_file:
                net.save_params(f_params=model_file)
```

Is this some code working in the background or is it simply the fact that the model parameters are stored in the `wandb_run_dir`?

borisdayma replied:
All files stored in wandb_run.dir are automatically saved.
You can see in my example run on the "files" tab

thomasjpfan replied:
Please leave a comment that states that the files in `wandb_run.dir` is automatically saved in `on_epoch_end`.

borisdayma commented:
Thanks for the comments @BenjaminBossan 
They should now be addressed but feel free to add anything.

I am using a new feature (`wandb.run.watch`) that has not been pushed to pypi yet so I'll update the requirements-dev with the correct version as soon as it's pushed.

BenjaminBossan commented:
@borisdayma Thank you for implementing the suggested changes. I will review this in detail when I have more time, hopefully soon.

borisdayma commented:
I just need to pin wandb version as soon as next release appear

borisdayma commented:
It should now be ready to be merged! Feel free if you have any questions or comments.

BenjaminBossan commented on skorch/callbacks/logging.py:
```diff
@@ -205,6 +207,82 @@ def on_train_end(self, net, **kwargs):
+      wandb Run used to log data.
+
+    save_model : bool (default=True)
+      Whether to save a checkpoint of the best model.
```
Maybe add an explanation here that the model will be uploaded to the W&B server for convenience.

borisdayma commented:
@BenjaminBossan Glad you found it useful!

I updated the docstring and added a comment for logging anonymously (without a W&B account). Feel free if you want me to change the wording or anything else.

thomasjpfan commented on skorch/callbacks/logging.py:
```diff
@@ -205,6 +207,86 @@ def on_train_end(self, net, **kwargs):
+    ... wandb_run = wandb.init(anonymous="allow)
+
+    >>> # Log hyper-parameters (optional)
+    ... wandb.config.update({"learning rate": 1e-3, "batch size": 32})
```
Is there a way to update this using the `wandb_run` object?

borisdayma replied:
Yes! I'll update

thomasjpfan commented on skorch/tests/callbacks/test_logging.py:
```diff
@@ -190,6 +191,101 @@ def test_first_batch_flag(
+        return mock
+
+    @pytest.fixture
+    def net_fitted(
```
This fixture is never used and can be removed.

thomasjpfan commented on skorch/callbacks/logging.py:
```diff
@@ -205,6 +207,86 @@ def on_train_end(self, net, **kwargs):
+        """Log values from the last history step and save best model"""
+        hist = net.history[-1]
+        keys_kept = filter_log_keys(hist, keys_ignored=self.keys_ignored_)
+        logged_vals = {k: hist[k] for k in keys_kept if k in hist}
```
Is this needed since `keys_kept` yields from `hist`?

```suggestion
        logged_vals = {k: hist[k] for k in keys_kept}
```

borisdayma commented:
All comments should have been implemented in last commit. Let me know if I missed anything.

borisdayma commented:
Just checking if you want me to add anything else to this PR

BenjaminBossan commented:
@borisdayma From my side, it looks good.

I had requested a re-review from @thomasjpfan since you addressed his comments. Maybe let's give him until the weekend to respond, if he doesn't we can consider it a thumbs up and merge.

BenjaminBossan commented:
Good job everyone, @borisdayma thanks for the contribution and your patience.

borisdayma commented:
Thanks, I think it will be very useful for fine-tuning. Feel free to reach if there's any issue related to it.

BenjaminBossan commented:
@borisdayma I now notice that ` conda install -c conda-forge --file requirements-dev.txt` fails with the following error:

```
PackagesNotFoundError: The following packages are not available from current channels:
  - wandb[version='>=0.8.30']
```

Is there a conda channel for wandb or is pip the only solution?

borisdayma commented:
At the moment it is only on pip but we can just use the new [conda interoperability with pip](https://docs.conda.io/projects/conda/en/latest/user-guide/configuration/pip-interoperability.html)

borisdayma commented:
Let me know if you want me to try it and suggest a new PR (to update docs).
I personally use pipenv all the time so I had not noticed this issue with conda.
